### ISSUE TITLE
How to train YOLOv3_tiny_3l on g4dn.12xlarge 4 GPU (NVIDIA T4) -> *** buffer overflow detected ***

### ISSUE THREAD

**Issue by aelich** [AlexeyAB/darknet#6400]
Dear,

When I trained **YOLOv3_tiny all was working** now tried to train **YOLOv3_tiny_3l and get error** 
```
*** buffer overflow detected ***
```
- I've compiled darknet with `GPU=1 CUDNN=1 CUDNN_HALF=1 OPENCV=1 DEBUG=0`
- I've changed `learning_rate=0.001` -> `learning_rate=0.00025`
- I've changed `burn_in=1000` -> `burn_in=4000`
- I've created weights `./darknet-cpu/darknet partial ../cfg/yolov3_tiny_training.cfg ../weights/yolov3-tiny.weights yolov3-tiny.conv.15 15`
- I've trained first 4000 iterations on 1GPU 
- then I turned on 4 GPU 

1GPU works
```
4502: 12.195651, 11.026999 avg loss, 0.000250 rate, 5.752831 seconds, 288128 images, 317.813969 hours left
Loaded: 0.000048 seconds
```

4GPU error after 42 iterations ERROR
```
Tensor Cores are disabled until the first 12000 iterations are reached.
 (next mAP calculation at 54926 iterations)
 4544: 25111717664524425734437759614976.000000, 2511171796675588063809505329152.000000 avg loss, 0.001000 rate, 1.811360 seconds, 1163264 images, 93.680821 hours left
*** buffer overflow detected ***: ./darknet terminated
Aborted (core dumped)
root@59812f3fcb3b:/home/darknet-user/deepleaningautomatization/TrainMe/3stepColabTraing#
```

My yolov3_tiny_3l.cfg
```
[net]
batch=64
subdivisions=1
loss_scale=128

learning_rate=0.00025
burn_in=4000
```
I trined up to 4000 weight on 1 GUP and later trea

```
[net]
# Testing
# batch=64
# subdivisions=4
# Training
batch=64
subdivisions=4
loss_scale=128
width=608
height=608
channels=3
momentum=0.9
decay=0.0005
angle=0
saturation = 1.5
exposure = 1.5
hue=.1

learning_rate=0.00025
burn_in=4000
max_batches = 216000
policy=steps
steps=172800,194400
scales=.1,.1

[convolutional]
batch_normalize=1
filters=16
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=32
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=64
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=1

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

###########

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=123
activation=linear



[yolo]
mask = 6,7,8
anchors = 6.9005,0.2301, 3.1682,0.7620, 0.7263,3.6008, 5.7680,0.5643, 0.4067,8.4283, 9.8830,0.4526, 10.6006,0.8410, 6.7824,1.4305, 1.1399,8.8201
classes=36
num=9
jitter=.3
ignore_thresh = .7
truth_thresh = 1
random=1

[route]
layers = -4

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[upsample]
stride=2

[route]
layers = -1, 8

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=123
activation=linear

[yolo]
mask = 3,4,5
anchors = 6.9005,0.2301, 3.1682,0.7620, 0.7263,3.6008, 5.7680,0.5643, 0.4067,8.4283, 9.8830,0.4526, 10.6006,0.8410, 6.7824,1.4305, 1.1399,8.8201
classes=36
num=9
jitter=.3
ignore_thresh = .7
truth_thresh = 1
random=1



[route]
layers = -3

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[upsample]
stride=2

[route]
layers = -1, 6

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=123
activation=linear

[yolo]
mask = 0,1,2
anchors = 6.9005,0.2301, 3.1682,0.7620, 0.7263,3.6008, 5.7680,0.5643, 0.4067,8.4283, 9.8830,0.4526, 10.6006,0.8410, 6.7824,1.4305, 1.1399,8.8201
classes=36
num=9
jitter=.3
ignore_thresh = .7
truth_thresh = 1
random=1
```

I've tried to lower `learning_rate=0.00025` to `learning_rate=0.00010` but it didnt help and average loss on 5000iterations stays near the same around 12.

Could you please help me @AlexeyAB  what am I doing wrong this approach worked on YOLOv3, YOLOv4, YOLOv3_tiny, YOLOv4_tiny altho with YOLOv4 I had different learning rate but also devided it by 4.

On YOLOv3_tiny_3l I always get same error **buffer overflow detected**

**Comment by AlexeyAB** [comment]
* Your anchors are wrong.

*  show screenshots with such information
```
./darknet detector test cfg/coco.data cfg/yolov4.cfg yolov4.weights data/dog.jpg
 CUDA-version: 10000 (10000), cuDNN: 7.4.2, CUDNN_HALF=1, GPU count: 1
 CUDNN_HALF=1
 OpenCV version: 4.2.0
 0 : compute_capability = 750, cudnn_half = 1, GPU: GeForce RTX 2070
net.optimized_memory = 0
mini_batch = 1, batch = 8, time_steps = 1, train = 0
   layer   filters  size/strd(dil)      input                output
   0 conv     32       3 x 3/ 1    608 x 608 x   3 ->  608 x 608 x  32 0.639 BF
```

**Comment by aelich** [comment]
- I calculated anchors using next command `./darknet detector calc_anchors data/obj.data -num_of_clusters 9 -width 19 -height 19` I put `19` as `608/32=19` and got values.  I set 9 clusters instead of 6 as I think yolov3-tiny-3l uses 9 maybe I need to leave just 2 decimal places instead of 4?

```
6.9005,0.2301, 3.1682,0.7620, 0.7263,3.6008, 5.7680,0.5643, 0.4067,8.4283, 9.8830,0.4526, 10.6006,0.8410, 6.7824,1.4305, 1.1399,8.8201
0.6749
```

- my screenshots is:

```
root@7417987f7407:/home/darknet-user/deepleaningautomatization/TrainMe/3stepColabTraing# ./5.continueTraining.sh 
 CUDA-version: 10020 (10020), cuDNN: 7.6.5, CUDNN_HALF=1, GPU count: 4  
 CUDNN_HALF=1 
 OpenCV version: 3.2.0
0,1,2,3
 Prepare additional network for mAP calculation...
 0 : compute_capability = 750, cudnn_half = 1, GPU: Tesla T4 
net.optimized_memory = 0 
mini_batch = 1, batch = 4, time_steps = 1, train = 0 
   layer   filters  size/strd(dil)      input                output
   0 conv     16       3 x 3/ 1    608 x 608 x   3 ->  608 x 608 x  16 0.319 BF

```

I cant find online what to do as 3l and 5l have really little info about and I'm learning now.

Thanks!

**Comment by AlexeyAB** [comment]
> I calculated anchors using next command ./darknet detector calc_anchors data/obj.data -num_of_clusters 9 -width 19 -height 19 I put 19 as 608/32=19 and got values. I set 9 clusters instead of 6 as I think yolov3-tiny-3l uses 9 maybe I need to leave just 2 decimal places instead of 4?
> Is this ok @AlexeyAB I done it to get same results as in script gen_anchors.py
> 
> 6.9005,0.2301, 3.1682,0.7620, 0.7263,3.6008, 5.7680,0.5643, 0.4067,8.4283, 9.8830,0.4526, 10.6006,0.8410, 6.7824,1.4305, 1.1399,8.8201
> 0.6749

This is for yolov2, not for yolov3/v4.

Use default anchors.


Show bad.list and bad_label.list

**Comment by aelich** [comment]
- I will train again with new anchors now will set to be int ./darknet detector calc_anchors data/obj.data -num_of_clusters 9 -width 608 -height 608 

Part that tricked me was in detector.c I saw you made gen_anchors.py into C code and where comparing values you got from `.py` and `.c` and I saw you didn't devide by 32 but I thought I should do it. 

darknet\src\detector.c (lines 1494 to 1496) 
```
    //gen_anchors.py = 1.19, 1.99, 2.79, 4.60, 4.53, 8.92, 8.06, 5.29, 10.32, 10.66
    //float orig_anch[] = { 1.19, 1.99, 2.79, 4.60, 4.53, 8.92, 8.06, 5.29, 10.32, 10.66 };
```
Yesterday I found you and **pjreddie** told it already

1. https://github.com/pjreddie/darknet/issues/555#issuecomment-376190325
2. https://github.com/pjreddie/darknet/issues/562#issuecomment-376202820

And in page 2 of [YOLOv3_whitepaper](https://arxiv.org/pdf/1804.02767.pdf) **pjreddie** 
```
"We still use k-means clustering to determine our bounding box priors. 
We just sort of chose 9 clusters and 3 scales arbitrarily and then divide up the clusters evenly across scales. 
On the COCO dataset the 9 clusters were:
(10×13),(16×30),(33×23),(30×61),(62×45),(59×119),(116 × 90),(156 × 198),(373 × 326)."
```

in [DARKNET](https://github.com/AlexeyAB/darknet)
```
Only if you are an expert in neural detection networks - recalculate anchors for your dataset for width and height from cfg-file: 
darknet.exe detector calc_anchors data/obj.data -num_of_clusters 9 -width 416 -height 416 
then set the same 9 anchors in each of 3 [yolo]-layers in your cfg-file. But you should change indexes of anchors masks= for each [yolo]-layer, 
so for YOLOv4 the 1st-[yolo]-layer has anchors smaller than 30x30, 2nd smaller than 60x60, 3rd remaining, and vice versa for YOLOv3. 
Also you should change the filters=(classes + 5)*<number of mask> before each [yolo]-layer. 
If many of the calculated anchors do not fit under the appropriate layers - then just try using all the default anchors.
```
and I get as YOLOv3_tiny works on two scales : 
```
[yolo]
mask = 3,4,5
...
[route]
layers = -4
[convolutional]
...
[upsample] <------(TO CHANGE SCALE - one pixel is transformed into a 4 pixels in a 2x2 area)
stride=2
...
[yolo]
mask = 0,1,2
```

and YOLOv3, YOLOv3_tiny_3l, YOLOv4... have 9 anchors and works on 3 scales
```
[yolo]
mask = 6,7,8
...
[upsample]
stride=2
...
[yolo]
mask = 3,4,6
...
[upsample]
stride=2
...
[yolo]
mask = 0,1,2
```

Stride I found in 

darknet\src\blas_kernels.cu (lines 1424 to 1429)
```
extern "C" void upsample_gpu(float *in, int w, int h, int c, int batch, int stride, int forward, float scale, float *out)
{
    size_t size = w*h*c*batch*stride*stride;
    upsample_kernel << <cuda_gridsize(size), BLOCK, 0, get_cuda_stream() >> >(size, in, w, h, c, batch, stride, forward, scale, out);
    CHECK_CUDA(cudaPeekAtLastError());
}
```
and darknet\src\blas.c
```
void upsample_cpu(float *in, int w, int h, int c, int batch, int stride, int forward, float scale, float *out)
{
    int i, j, k, b;
    for (b = 0; b < batch; ++b) {
        for (k = 0; k < c; ++k) {
            for (j = 0; j < h*stride; ++j) {
                for (i = 0; i < w*stride; ++i) {
                    int in_index = b*w*h*c + k*w*h + (j / stride)*w + i / stride;
                    int out_index = b*w*h*c*stride*stride + k*w*h*stride*stride + j*w*stride + i;
                    if (forward) out[out_index] = scale*in[in_index];
                    else in[in_index] += scale*out[out_index];
                }
            }
        }
    }
}
```
so I saw `h*stride, w*stride` so for `stride=2` in `[upsample]` it makes it double bigger.

I'm just so **confused** why on smaller layer we use mask=3,4,5 and anchors with bigger values and then after `[upsample] stride=2` we use mask=0,1,2 and anchors smaller than 30x30. 

**_As masks determent which anchors will be used shouldn't be opposite_**???

@AlexeyAB please sorry for boring you with all of these questions.

**Comment by aelich** [comment]
- I don't have these files `bad.list` or `bad_label.list` my darknet isnt generating them darknet\src\data.c lines 

bad.list (lines 201 to 216)
```
if (!file) {
        printf("Can't open label file. (This can be normal only if you use MSCOCO): %s \n", filename);
        //file_error(filename);
        FILE* fw = fopen("bad.list", "a");
        fwrite(filename, sizeof(char), strlen(filename), fw);
        char *new_line = "\n";
        fwrite(new_line, sizeof(char), strlen(new_line), fw);
        fclose(fw);
        if (check_mistakes) {
            printf("\n Error in read_boxes() \n");
            getchar();
        }

        *n = 0;
        return boxes;
    }
```
bad_label.list (408 to 415)
```
if (id >= classes) {
            printf("\n Wrong annotation: class_id = %d. But class_id should be [from 0 to %d], file: %s \n", id, (classes-1), labelpath);
            sprintf(buff, "echo %s \"Wrong annotation: class_id = %d. But class_id should be [from 0 to %d]\" >> bad_label.list", labelpath, id, (classes-1));
            system(buff);
            if (check_mistakes) getchar();
            ++sub;
            continue;
        }
```


- I **have really difficult dataset** bills from USA [Data Sample](https://www.pplelectric.com/-/media/PPLElectric/My-Account/Images/Sample-bill-page-1-06-09-15.gif) here I need to detect **Amount due** and all bbox are really long and thin. So **default anchors** gave bad me depressingly bad results when I used them sometime ago.

My dataset in 600k images in size and I can detect all but boxes drawn are often moved or cut some parts of data and that's why I committed to fix anchors. I have rotation for 90 - 90 degrees  my anchors are ugly **_because boxes are all long and thin._**:
[220, 7, 101, 24, 23, 115, 184, 18, 13, 269, 316, 14, 339, 26, 217, 45, 36, 282]

So i have huge problem of sorting them and correlate to [upsample] like you and **pjreddie** done before.

@AlexeyAB any advice would be deeply appreciated.

**Comment by AlexeyAB** [comment]
https://github.com/AlexeyAB/darknet#how-to-improve-object-detection

> Only if you are an expert in neural detection networks - recalculate anchors for your dataset for width and height from cfg-file: darknet.exe detector calc_anchors data/obj.data -num_of_clusters 9 -width 416 -height 416 then set the same 9 anchors in each of 3 [yolo]-layers in your cfg-file. But you should change indexes of anchors masks= for each [yolo]-layer, so for YOLOv4 the 1st-[yolo]-layer has anchors smaller than 30x30, 2nd smaller than 60x60, 3rd remaining, and vice versa for YOLOv3. Also you should change the filters=(classes + 5)*<number of mask> before each [yolo]-layer. If many of the calculated anchors do not fit under the appropriate layers - then just try using all the default anchors.
